// src/sockets/socket.server.js
const { Server } = require("socket.io");
const cookie = require('cookie');
const jwt = require('jsonwebtoken')
const userModel = require("../models/user.model")
const aiService = require("../services/ai.service")
const messageModel = require("../models/message.model")
const { createMemory, queryMemory } = require("../services/vector.service")

function initSocketServer(httpServer) {
    const io = new Server(httpServer, {})


    io.use(async (socket, next) => {
        const cookies = cookie.parse(socket.handshake.headers?.cookie || "");


        if(!cookies.token) {
            next(new Error("Authenticated error: No token provided"));

        }

        try{
            const decoded = jwt.verify(cookies.token, process.env.JWT_SECRET);

            const user = await userModel.findById(decoded.id);

            socket.user = user

            next()



        }catch(err) {
            next(new Error("Authenticated error: Invalid token"));
        }
    })


    io.on("connection", (socket) => {
       
        socket.on("ai-message", async(messagePayload)=> {
            


            const [ usermessage, vectors ] = await Promise.all([
                messageModel.create({
                chat: messagePayload.chat,
                user: socket.user._id,
                content: messagePayload.content,
                role: "user"
            }),
            aiService.generateVector(messagePayload.content),
            ])

            await createMemory({
                vectors,
                messageID: usermessage._id,
                text: messagePayload.content,
                metadata: {
                    chat: messagePayload.chat.toString(),
                    user: socket.user._id.toString(),
                    role: "user"
                }
            })


            

            const [ memory, chatHistory ] = await Promise.all([
                queryMemory({
                queryVector: vectors,
                limit: 3,
                
            }),
            messageModel.find({
                chat: messagePayload.chat
            }).sort({createdAt: -1}).limit(8).lean()

            ]);


            chatHistory.reverse();


            //data(chathistory) fetch karenge database se (short term memory)
            
            
            const stm = chatHistory.map(item => ({
                role: item.role,
                content: item.content
            }));
            
            const ltm = {
                role: "system",
                content: `${memory.slice(0,3).map(item => item.metadata.text).join("\n")}`
            };


            console.log(ltm)
            console.log(stm)

            // generate response from AI
            const response = await aiService.generateResponse([ ltm, ...stm ]);


            
            socket.emit('ai-response', {
                content: response,
                chat: messagePayload.chat
            })

           // ai ka respone database mein save karenge
            const responseMessage = await messageModel.create({
                chat: messagePayload.chat,
                user: socket.user._id,
                content: response,
                role: "assistant"
            });

            

            // ai ka response vector database mein save karenge
            await createMemory({
                messageID: responseMessage._id,
                text: response,
                metadata: {
                    chat: messagePayload.chat.toString(),
                    user: socket.user._id.toString(),
                    role: "assistant" 
                }
            });


            


        })
    })


}


module.exports = initSocketServer;
