// src/services/ai.service.js
const OpenAI = require("openai");
const { GoogleGenAI } = require("@google/genai");


console.log("Gemini Key:", !!process.env.GEMINI_API_KEY);

// -------- GROK (chat / response) --------
const grokClient = new OpenAI({
  apiKey: process.env.GROQ_API_KEY,
  baseURL: "https://api.groq.com/openai/v1",
});

// -------- GEMINI (embeddings only) --------
const ai = new GoogleGenAI({
  apiKey: process.env.GEMINI_API_KEY
});


// Grok response
async function generateResponse(messages) {
  // Convert messages array → single string
  const prompt = messages
    .map(m => `${m.role}: ${m.content}`)
    .join("\n")
    .slice(-12000);

  const response = await grokClient.responses.create({
    model: "openai/gpt-oss-20b",
    input: prompt
  });

  return response.output_text;
}



// Gemini embeddings (CORRECT & WORKING)
async function generateVector(content) {
  const response = await ai.models.embedContent({
    model: "gemini-embedding-001",
    contents: content,
    config: {
      outputDimensionality:768 
    }
  });

  return response.embeddings[0].values;
}

module.exports = {
  generateResponse,
  generateVector
};
